### PR TITLE
Release/v0.1.0

### PR DESCRIPTION
**Pull request by jdevcs**
- Main project is now template CLI project 

- There are two templates ( Hardhat+ Web3.js v4 + JS + NPM + Token contract and its tests) 
- and ( Hardhat+ Web3.js v4 + TS + Yarn +  Lock contract plus its tests)

- After installing this package and with `npx create-hardhat-web3 init js` or `npx create-hardhat-web3 init ts` users can create project using template

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,17 +26,37 @@ Follow these steps to set up and run the project locally.
...
 ### Installation
 
-1. Clone the repository:
+1. Make a new directory:
    
-   `git clone git@github.com:web3/hardhat-web3.git`
+   `mkdir newproj`
 
-2. Install NPM packages:
+2. Create node project using NPm or Yarn:
```
Comment (luu-alex) [inline README.md:33]:
```suggestion
2. Create node project using NPM or Yarn:
```

File: template/js-template/README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+This project demonstrates a basic Hardhat use case, focusing on testing a Token smart contract. It includes a sample contract, a test for that contract, and a script that deploys the contract.
+
+The project utilizes:
+- Hardhat as the development environment
+- Chai for assertions
+- web3.js for interacting with Ethereum
+
+## Prerequisites
+
+- [Node.js and npm](https://nodejs.org/en/download/) (latest version)
```
Comment (luu-alex) [inline template/js-template/README.md:17]:
```suggestion
- [Node.js and NPM](https://nodejs.org/en/download/) (latest version)
```

File: template/js-template/README.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+ Install NPM packages:
+   
+   `npm install`
+
+### Compiling the Contract
+
+Compile the smart contract with:
+
+`npx hardhat compile`
```
Comment (luu-alex) [inline template/js-template/README.md:34]: when running this command i get an error
```
➜  js-template git:(release/v0.1.0) ✗ npx hardhat compile
Error HH12: Trying to use a non-local installation of Hardhat, which is not supported.
Please install Hardhat locally using npm or Yarn, and try again.

For more info go to https://hardhat.org/HH12 or run Hardhat with --show-stack-traces
```

File: template/ts-template/test/Lock.ts [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    // use first default account if not provided explicitly for sending transaction
+    if (web3.utils.isNullish(txOptionsObj.from))
+      txOptionsObj = {
+        from: (await web3.eth.getAccounts())[0],
+        ...txOptionsObj
+      };
+
+    const artifact: Artifact = await hre.artifacts.readArtifact(contractName);
+
+    //for exact types support we will use @chainsafe/hardhat-ts-artifact-plugin by npx hardhat ts-artifact after npx hardhat compile and pass that to abi function param
```
Comment (luu-alex) [inline template/ts-template/test/Lock.ts:61]:
```suggestion
    // For exact types support we will use @chainsafe/hardhat-ts-artifact-plugin by npx hardhat ts-artifact after npx hardhat compile and pass that to abi function param
```

File: template/js-template/README.md (hunk 3 above)
Comment (jdevcs) [reply template/js-template/README.md:34]: before `npx hardhat compile`,  install dependencies with `yarn` or `npm i`

**Review by danforbes — COMMENTED** [review]
Is there a reason why this script requires the user to init a new NPM project and install dependencies, etc? For instance, the ZKsync CLI creates a new directory, initializes the NPM project and GitHub repo, etc. That pattern may be more familiar to our users.

**Comment by jdevcs** [comment]
> Is there a reason why this script requires the user to init a new NPM project and install dependencies, etc? For instance, the ZKsync CLI creates a new directory, initializes the NPM project and GitHub repo, etc. That pattern may be more familiar to our users.

That is also possible:
  install above package in global scope (as its not published yet so for local testing `yarn link`) 
  and after that in any new dir: `npx create-hardhat-web3 init ts ` or `npx create-hardhat-web3 init js `

File: template/ts-template/test/Lock.ts [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    const artifact: Artifact = await hre.artifacts.readArtifact(contractName);
+
+    const contract = new web3.eth.Contract(artifact.abi);  // this will not have full types support 
+
+    const deployedContract = await contract.deploy({
+      data: artifact.bytecode,
+      arguments: constructorArgs
+    }).send(txOptionsObj);
+
+    deployedContract.defaultAccount = txOptionsObj.from; // this will be used for sending future transactions to contract for all .send() function calls
```
Comment (danforbes) [inline template/ts-template/test/Lock.ts:39]: This comment may be a little difficult for people to read. Personally, I prefer comments that appear on the line above a piece of code as opposed to comments that appear on the same line. I think especially for longer comments like this it's nice to use a multi-line comment above the code that's being commented.

File: template/ts-template/test/Lock.ts [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    const deployedContract = await contract.deploy({
+      data: artifact.bytecode,
+      arguments: constructorArgs
+    }).send(txOptionsObj);
+
+    deployedContract.defaultAccount = txOptionsObj.from; // this will be used for sending future transactions to contract for all .send() function calls 
+    return deployedContract;
+  }
+
+  //TODO: add  @chainsafe/hardhat-ts-artifact-plugin output support in Artifacts object in HRE and use above function for reading abi from there with full types support
```
Comment (danforbes) [inline template/ts-template/test/Lock.ts:43]: Similar to comment above...
```suggestion
  // TODO: add  @chainsafe/hardhat-ts-artifact-plugin output support in
  // Artifacts object in HRE and use above function for reading abi from
  // there with full types support
```

File: bin/init-script.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  fs.copy(sourceDir, targetDir)
+    .then(() => {
+      console.log(`Project initialized successfully in ${targetDir} using ${language === 'js' ? 'JavaScript' : 'TypeScript'} template.`);
+    })
+    .catch((err) => {
+      console.error('Error initializing project:', err);
+    });
+} else {
+  console.log('Unknown command. Use "init [js|ts]" to create a new project.');
+}
```
Comment (danforbes) [inline bin/init-script.js:35]: Nitpick: end a newline to the end of the file.
```suggestion
}

```

**Review by luu-alex — APPROVED** [review]

File: template/ts-template/test/Lock.ts [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+      lock,
+      unlockTime,
+      lockedAmount,
+      owner,
+      otherAccount,
+      web3
+    };
+  }
+
+  describe("Deployment", function () {
```
Comment (danforbes) [inline template/ts-template/test/Lock.ts:97]: Is there a reason why the tests are not in separate files?

File: template/ts-template/test/Lock.ts [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+          withdrawalEventPromise
+        ]);
+
+        expect(gotLockedAmount).to.be.equal(BigInt(lockedAmount))
+        expect(result.status).to.equal(1n);
+
+      });
+    });
+  });
+});
```
Comment (danforbes) [inline template/ts-template/test/Lock.ts:220]:
```suggestion
});

```

**Review by danforbes — APPROVED** [review]
Very well documented! Mainly formatting comments and ease-of-use comments from me.

File: template/js-template/README.md (hunk 3 above)
Comment (luu-alex) [reply template/js-template/README.md:34]: worked. thanks

Comment (jdevcs) [reply template/js-template/README.md:34]: > What is the reasoning for not running these commands as part of the init script?

so users are aware of steps usually required for building proj.

File: README.md [hunk 10]
```diff
@@ -28,17 +26,37 @@ Follow these steps to set up and run the project locally.
...
+
+3. Install create-hardhat-web3 .
+
+   `npm i create-hardhat-web3` or
+   `yarn add create-hardhat-web3`
+
+4. Use required template for initilization of your project.
+
+For Hardhat, Web3.js, Javascript project :
+   `npx create-hardhat-web3 init` or
```
[the remark this answers is not in the document]
Comment (jdevcs) [reply README.md:46]: > Will the script work without the language being specified?

no, user will need to specify js or ts

File: template/ts-template/test/Lock.ts (hunk 8 above)
Comment (jdevcs) [reply template/ts-template/test/Lock.ts:97]: Currently these are grouped in a file by contract name and using fixtures from same file, just for simplicity for starting.

**Comment by jdevcs** [comment]
made above formatting changes and published https://www.npmjs.com/package/create-hardhat-web3

**Comment by jdevcs** [comment]
@danforbes For further ideas and improvements could you create new issues. Thanks

**Pull request merged**